Handle shadow host with delegatesFocus=true in focusing steps (click focus + focus())

rakina opened the pull request:
- [x] At least two implementers are interested (and none opposed):
   * Blink
   * Gecko & Safari are reviewing the other PRs, so marking as interested
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   *  https://github.com/web-platform-tests/wpt/pull/18035
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: TODO
   * Firefox: TODO
   * Safari: TODO

Another part of #2013 and whatwg/dom#768. Defines the behavior of focus() and the focusing steps in general.
Note that this is a change made on top of https://github.com/whatwg/html/pull/4735, which defines flattened tabindex-ordered focus navigation scopes, etc. needed for this change.

Another note, the behavior in Chrome for focus() when the flattened tabindex-ordered focus navigation scope is empty (there's nothing to focus on) is kinda weird - it focuses on the host instead of not focusing on anything ([crbug](https://bugs.chromium.org/p/chromium/issues/detail?id=978299&q=reporter%3Ame%20&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified) filed). I'm not sure if we want this behavior or not in the specs (the current PR aborts the focusing steps instead).


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/4796/browsing-the-web.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">/browsing-the-web.html</a>  ( <a href="https://whatpr.org/html/4796/562631f...f4f734a/browsing-the-web.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">diff</a> )
<a href="https://whatpr.org/html/4796/infrastructure.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/4796/562631f...f4f734a/infrastructure.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">diff</a> )
<a href="https://whatpr.org/html/4796/interaction.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">/interaction.html</a>  ( <a href="https://whatpr.org/html/4796/562631f...f4f734a/interaction.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">diff</a> )
<a href="https://whatpr.org/html/4796/references.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">/references.html</a>  ( <a href="https://whatpr.org/html/4796/562631f...f4f734a/references.html" title="Last updated on Oct 16, 2019, 6:40 AM UTC (f4f734a)">diff</a> )

rakina commented:
cc people involved in previous PRs @rniwa @annevk @domenic

rniwa commented on source:
```diff
@@ -73547,8 +73553,92 @@ END:VCARD</pre>
+   data-x="concept-DocumentFragment-host">host</span>, then return <var>element</var>'s
+   <span>assigned slot</span>.</p></li>
+
+   <li><p>If <var>element</var>'s parent is its <span>shadow root</span>, then return the parent's
```
"its shadow root" in "If element's parent is its shadow root" is ambiguous here. I think what you meant is that element's parent is a shadow root which is the root of the element? It's unclear what its shadow root means in this context. I first thought maybe the element itself is a shadow host but it's impossible for element's parent to that shadow root so I got confused.

rakina replied:
Thanks, going to change this on the sequential focus navigation PR. Since this PR is actually just this commit:https://github.com/whatwg/html/pull/4796/commits/66f6b5e6360ddf8872ace5c337cb1a6fa5a9914f (made on top of the sequential focus navigation PR), can you add these comments to https://github.com/whatwg/html/pull/4735 instead?

rniwa commented on source:
```diff
@@ -73547,8 +73553,92 @@ END:VCARD</pre>
+  <ul>
+   <li><p>It contains all elements in <var>owner</var>'s <span>focus navigation scope</span> that
+   are themselves <span data-x="focus navigation scope owner">focus navigation scope owners</span>,
+   except the elements whose <span>tabindex value</span> is a negative integer.</p></li>
```
"It" is ambiguous. Is it referring to tabindex-ordered focus navigation scope? We should explicitly state the term instead. This is a very roundabout way of defining the scope. I think a more clear way to define this would be to define the scope of a given element. Instead of saying a scope S contains an element X, say element X's scope is S.

rakina replied:
(as mentioned above, this PR is actually just this commit:66f6b5e made on top of the sequential focus navigation PR, can you add these comments to #4735 instead?)

rniwa commented on source:
```diff
@@ -73547,8 +73553,92 @@ END:VCARD</pre>
+  <p>A <dfn>flattened tabindex-ordered focus navigation scope</dfn> is a list of <span
+  data-x="focusable area">focusable areas</span>. Every <span>focus navigation scope owner</span>
+  <var>owner</var> owns a distinct <span>flattened tabindex-ordered focus navigation scope</span>,
+  whose contents are determined by the following algorithm:</p>
```
Ugh... similar to the comments I left for sequential tab index order for shadow trees, I really hate the approach of merging lists of lists. It just makes it impossible to reason & think about given an element X, what the next element comes in the tab index order.

domenic commented:
Also let's call special attention to @rakina's note in the OP:

> Another note, the behavior in Chrome for focus() when the flattened tabindex-ordered focus navigation scope is empty (there's nothing to focus on) is kinda weird - it focuses on the host instead of not focusing on anything (crbug filed). I'm not sure if we want this behavior or not in the specs (the current PR aborts the focusing steps instead).

It sounds like Chrome would prefer to abort focusing in such cases (e.g. delegatesFocus on a shadow host that has no children). Does that make sense to other browsers? We can certainly iterate on it later, similar to https://github.com/whatwg/html/pull/4731#issuecomment-516228512 for `:focus`, but it's also nice to get folks impressions.

rniwa commented on source:
```diff
@@ -73916,6 +74025,17 @@ END:VCARD</pre>
 
+     <dt>If <var>new focus target</var> is a <span>shadow host</span> whose <span
+     data-x="concept-element-shadow-root">shadow root</span>'s <span>delegates focus</span> is true
+     and whose <span>flattened tabindex-ordered focus navigation scope</span> <var>scope</var> is
```
Hm... this might be a bit problematic for Safari where certain elements are not focusable for the purpose of sequential navigation with tab key but it is with option+tab. Which one of the list are using here?

rakina replied:
Hmm, isn't that generally true for every sequential-focus-related parts of the spec though? I'm not sure how we can say that different methods of sequential navigation may yield different results.

rniwa replied:
This is particularly problematic because it gets triggered by JS API: `element.focus()` whereas the regular sequential navigation happens via the user input.

rakina replied:
Hmmm, I see... I'm not sure about the use cases of skipping some elements in Safari, but since this is not actually sequential navigation, can we just include *all* elements whose tabindex >= 0 and choose the first one in this case?

rniwa replied:
The question is whether an anchor element without tabindex, etc... are focused. If not, and we're only considering the tab-index value, that's really a new kind of algorithm / behavior which needs to be defined somewhere.

We probably should do that because this is a JS API, and any sort of non-interoperability across browsers are gonna result in a developer pain.

domenic replied:
Interesting. I think I see the problem.

This spec text insertion is is the focusing steps, which are used whenever the UA has made an explicit decision to focus something. This can occur in several situations:

- Programmatic focus, e.g. `autofocus=""` or `element.focus()`.
- During sequential focus navigation, _if the UA has determined that the element is sequentially focusable_
- As a result of clicking, _if the UA has determined that the element is click focusable_.

For programmatic focus, this insertion works fine.

However, during sequential focus navigation or click focusing:

- The UA might determine whether the element is sequentially/click focusable depending on the sequential/click focusability of elements in _scope_. This is fine with this PR. If it determines that it is not, then it will not call these focusing steps.
- The UA might want to set _new focus target_ to a specific element in _scope_, that is the first sequentially/click focusable area (according to the UA's definitions). This does not work well with this PR, which explicitly calls for the first focusable area of _scope_, even if it is not sequentially/click focusable.

I think an appropriate resolution here would be to give the UA discretion in choosing, when the focusing steps are being run as a result of sequential focus navigation or click focusing. (Concepts defined in #4768.)

We could make that fully rigorous by threading some kind of "cause" argument to the focusing steps (which would be either programmatic, sequential, or click), and then giving discretion when the cause argument is sequential or click. We could also hand-wave a bit and just talk about "when this algorithm is called as a result of sequential focus navigation" etc.

Thoughts?

rniwa replied:
So the idea is to have the algorithm take some kind of argument and use programmatic focus mode here?

Note that my main concern here is that we need to make the behavior of `Element.focus()` same across all browses when called on a shadow root with *delegates focus* flag set.

I guess you're saying that the issue is more to do with the current algorithm not giving discretion for UA to decide what's sequentially focusable?

domenic replied:
> So the idea is to have the algorithm take some kind of argument and use programmatic focus mode here?

Not quite. My idea was that the algorithm takes an argument determining what mode it's in (sequential, click, or programmatic), and then uses that mode to decide which member of _scope_ to focus, instead of always focusing the first focusable one.

> Note that my main concern here is that we need to make the behavior of Element.focus() same across all browses when called on a shadow root with delegates focus flag set.

I think that is taken care of by the current PR. It should always be the first focusable (i.e., programmatically focusable) child in _scope_ (which will always exist). It's the non-programmatic cases that are trickier.

> I guess you're saying that the issue is more to do with the current algorithm not giving discretion for UA to decide what's sequentially focusable?

Sort of. In particular the current algorithm assumes the UA has already applied that discretion, before deciding whether and on what to call the algorithm. The call sites of this algorithm specifically allow such discretion.

However, now that we have this sub-scopes situation, we need to insert a second opportunity for UA discretion to apply.

rakina replied:
> I think that is taken care of by the current PR. It should always be the first focusable (i.e., programmatically focusable) child in scope (which will always exist). It's the non-programmatic cases that are trickier.

I might misunderstood what you mean here - but, it's actually the first "sequentially" focusable child in scope (e.g. those with tabindex=-1 won't be focused). So like @rniwa said, it might be different if we actually use the same tabindex-ordered scope as the one for sequential navigation.

domenic replied:
Oh dang OK, my bad for misreading, and thanks for catching me.

The best outcome here then, IMO, is:

- We define a counterpart of "flattened tabindex-ordered focus navigation scope" which is different in two ways: it uses DOM order, and it does not omit negative tabindexes. Call it "flattened DOM-ordered focus navigation scope" maybe?
- We add a type-of-focus flag to the focusing steps algorithm.
- We pass that flag correctly from the various callers. (Maybe it can default to programmatic.)
- At this step, "If new focus target is a shadow host whose shadow root's delegates focus is true", then we switch on the type-of-focus:
  - If programmatic, pick the first focusable area in the "flattened DOM-ordered focus navigation scope"
  - If sequential, pick the first focusable area in the flattened tabindex-ordered focus navigation scope that is sequentially focusable.
  - If click, pick the first focusable area in the flattened DOM-ordered focus navigation scope that is click focusable.

Do you think this is reasonable? It sounds like it doesn't match what Chrome does, and it is definitely more complex. But I think the model in this PR, which uses tabindex-ordering even for programmatic focus, is pretty weird...

rakina replied:
Sorry, I should've been clearer - the order is also the same as sequentially focusable (will focus on tabindex=1 element instead of tabindex=0 even if the latter comes first in DOM order) - see test: https://github.com/web-platform-tests/wpt/pull/18035/files#diff-4c70a5c2994792409f0e8bbe698ae3feR206

Let me propose something simpler - please point out if I'm missing some cases:
We define flattened tabindex-ordered focus navigation scope to contain *all* focusable areas whose tabindex value is non-negative.
* Sequential focus navigation will still work because we will only focus on suitable sequentially focusable areas, which considers user agent quirks (defined here https://whatpr.org/html/4768/2a397f6...95bb810/interaction.html#suitable-sequentially-focusable-area)
* Focusing steps for clicks will only run for click focusable focusable areas (https://whatpr.org/html/4768/2a397f6...95bb810/interaction.html#click-focusable)
* In the focusing steps, for a focus target with delegatesFocus=true, we just use flattened tabindex-ordered focus navigation scope - since its contents doesn't care about user agent quirks, all user agents will yield the same result.


I'll rebase this patch on top of https://github.com/whatwg/html/pull/4768 if that sounds good!

rakina replied:
Did you mean for the span to have`tabindex=1` instead of `-1`? Because otherwise the actual result (in Chrome as implemented now, at least) will never be the span because elements with negative tabindex not included in the flattened tabindex-ordered focus navigation scope.

If you actually mean `tabindex=1`, then the results will be just like you listed. I think my proposal does work, no? both click and `.focus()`will result in the span getting focused because they're purely tabindex-based. Sequential focus will skip the checkbox depending on the os + keys because it needs to find the next "suitable sequentially focusable area".

rakina replied:
Aaah ok I finally get it, thanks and sorry D:
I agree that using sequential navigation for this is weird, and I think your previous proposal is reasonable.  From the thread and explainer of delegatesFocus (https://github.com/w3c/webcomponents/issues/126, https://github.com/w3c/webcomponents/blob/gh-pages/proposals/ShadowRoot-delegatesFocus-Proposal.md) there seems to be no specific reason to use sequential focus navigation too.

I'm a bit worried about potential compatibility issues for things that already depend on Chrome's behavior though, but focus is kinda obscure and underspec-ed so maybe that number is very low?

rakina replied:
Filed https://github.com/w3c/webcomponents/issues/830. Thank you!

smaug---- replied:
I agree, merging PRs when big picture isn't clear should be avoided.

rniwa commented:
> > Another note, the behavior in Chrome for focus() when the flattened tabindex-ordered focus navigation scope is empty (there's nothing to focus on) is kinda weird - it focuses on the host instead of not focusing on anything (crbug filed). I'm not sure if we want this behavior or not in the specs (the current PR aborts the focusing steps instead).
> 
> It sounds like Chrome would prefer to abort focusing in such cases (e.g. delegatesFocus on a shadow host that has no children). Does that make sense to other browsers? We can certainly iterate on it later, similar to [#4731 (comment)](https://github.com/whatwg/html/pull/4731#issuecomment-516228512) for `:focus`, but it's also nice to get folks impressions.

We should do whatever matches when focusing an element that's not focusable, which is basically aborting.

rakina commented:
> > > Another note, the behavior in Chrome for focus() when the flattened tabindex-ordered focus navigation scope is empty (there's nothing to focus on) is kinda weird - it focuses on the host instead of not focusing on anything (crbug filed). I'm not sure if we want this behavior or not in the specs (the current PR aborts the focusing steps instead).
> > 
> > 
> > It sounds like Chrome would prefer to abort focusing in such cases (e.g. delegatesFocus on a shadow host that has no children). Does that make sense to other browsers? We can certainly iterate on it later, similar to [#4731 (comment)](https://github.com/whatwg/html/pull/4731#issuecomment-516228512) for `:focus`, but it's also nice to get folks impressions.
> 
> We should do whatever matches when focusing an element that's not focusable, which is basically aborting.

SGTM! This matchecs the spec PR then.

rakina commented:
> This isn't really reviewable since the patch has somehow #4735 in it.

Here's the changes on top of that PR: https://github.com/whatwg/html/pull/4796/commits/99c8181ef8fb8b67c4ac0f192ba0f5f69de443d4

rakina commented:
Per https://github.com/w3c/webcomponents/issues/830#issuecomment-532514644, this needs an update to use the flat tree order and also use the "click focusable" definition from https://github.com/whatwg/html/pull/4768. I'll update this next week - in the meantime, @smaug---- and @rniwa  can you review and maybe give approvals on https://github.com/whatwg/html/pull/4735 and https://github.com/whatwg/html/pull/4768/files/4e2a998af84570fda3ff90d45a1882408c9f1cbb..75b444334fe8a77d6e26ee8b4c160ea6cebf3d93?

rakina commented:
Rebased and updated the PR to use first in flat tree order (+ click focusable for focus). Note that sequential navigation focus delegation already works fine because we already skip hosts with delegatesFocus=true in sequential navigation.

annevk commented:
There's a couple things I don't quite understand:

1. Why do we still reference the box tree for some of this? Wouldn't the box tree also cross shadow tree boundaries, including those without delegate focus? I suspect we should update those to say flat tree as well with the appropriate checks (if flat tree allows for those checks, haven't checked).
2. How does this deal with recursion? If a shadow tree contains another shadow tree without delegate focus, it seems that only looking at the flat tree would blow straight past that.

emilio commented:
Yeah, using the flat tree directly to dig down in the tree just ignores any other potential shadow tree which may have delegateFocus = false, I think... I think looking at flattened tree ancestors may be ok here, but downwards you get into funny behavior.

What Gecko implements for focus navigation in shadow trees (and as far as I can tell, that also matches Blink and WebKit) is mostly http://w3c.github.io/webcomponents/spec/shadow/#sequential-focus-navigation, why do we need to come up with yet a different algorithm for this case?

The code for focusing an element via `focus()` is https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/dom/base/nsFocusManager.cpp#1140

The code for click focus handling is here: https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/dom/events/EventStateManager.cpp#3035

And effectively just checks whether the event hasn't been prevented, and looks at focusability of the event target or any flattened tree ancestor.

rakina commented:
@emilio:
See this thread on why we want a different behavior, it was also talked
about in TPAC:
https://github.com/w3c/webcomponents/issues/830#issuecomment-532514644

On Thu, Sep 26, 2019, 19:16 Emilio Cobos Álvarez <notifications@github.com>
wrote:

> Yeah, using the flat tree directly to dig down in the tree just ignores
> any other potential shadow tree which may have delegateFocus = false, I
> think... I think looking at flattened tree ancestors may be ok here, but
> downwards you get into funny behavior.
>
> What Gecko implements for focus navigation in shadow trees (and as far as
> I can tell, that also matches Blink and WebKit) is mostly
> http://w3c.github.io/webcomponents/spec/shadow/#sequential-focus-navigation,
> why do we need to come up with yet a different algorithm for this case?
>
> The code for focusing an element via focus() is
> https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/dom/base/nsFocusManager.cpp#1140
>
> The code for click focus handling is here:
> https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/dom/events/EventStateManager.cpp#3035
>
> And effectively just checks whether the event hasn't been prevented, and
> looks at focusability of the event target or any flattened tree ancestor.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/whatwg/html/pull/4796?email_source=notifications&email_token=ABPN2EVFPI6PKE5HNPZBR6TQLSDXFA5CNFSM4IGP2CK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7VCHSA#issuecomment-535438280>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN2EVJNH7JW66XFDGGS7LQLSDXFANCNFSM4IGP2CKQ>
> .
>

rakina commented:
> Why do we still reference the box tree for some of this? Wouldn't the box tree also cross shadow tree boundaries, including those without delegate focus? I suspect we should update those to say flat tree as well with the appropriate checks (if flat tree allows for those checks, haven't checked).

I don't think I understand this, which part are you referring to?

> How does this deal with recursion? If a shadow tree contains another shadow tree without delegate focus, it seems that only looking at the flat tree would blow straight past that.

I think this is OK. "delegates focus" doesn't mean "you now can focus on things inside my shadow tree", it only means "don't focus on the host". So even if a child shadow tree has `delegatesFocus=false` it doesn't mean "don't put focus on things inside", so there should be no problem of just using the flat tree order? As discussed in https://github.com/w3c/webcomponents/issues/830, this behavior makes the focus delegation predictable and interoperable.

rakina commented:
> Per discussion in [web-platform-tests/wpt#18035 (comment)](https://github.com/web-platform-tests/wpt/pull/18035#issuecomment-535424532) we need to not use activation behavior here, but instead something like "when a user activates them". For now I think the best cross-reference for "activates" is just an `<a href="#activation">`. Eventually we may get better explanation for the end-to-end process of users activating things, by integrating with https://w3c.github.io/uievents/#event-flow-activation, but for now that seems good enough.
> 
> We may want to add an explicit `<p class="note">` stating that focusing is not activation behavior, i.e. dispatching synthetic `click` events or calling `click()` will not cause focus.

Added, thanks!

annevk commented:
@rakina ah, thanks, and there's no chance of a descendant host getting focus while it should not?

As for "box tree", it's mentioned twice in HTML's focus sections and it seems somewhat redundant with "flat tree".

rakina commented:
> @rakina <https://github.com/rakina> ah, thanks, and there's no chance of
a descendant host getting focus while it should not?


Yes, because hosts with delegatesFocus=true are not considered as focusable
areas, so they'll never be in the "possible focus delegates" list.


> As for "box tree", it's mentioned twice in HTML's focus sections and it
seems somewhat redundant with "flat tree".


I see, you're right - I updated the PR to remove the ordering part and
change box tree to flat tree in the other one. Thanks!


On Fri, Sep 27, 2019, 07:36 Anne van Kesteren <notifications@github.com>
wrote:

> @rakina <https://github.com/rakina> ah, thanks, and there's no chance of
> a descendant host getting focus while it should not?
>
> As for "box tree", it's mentioned twice in HTML's focus sections and it
> seems somewhat redundant with "flat tree".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/whatwg/html/pull/4796?email_source=notifications&email_token=ABPN2EXYXUZRK57BJLKUZQTQLXV2VA5CNFSM4IGP2CK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7YTXYQ#issuecomment-535903202>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN2ERFHXTMLK6ZCFE4IK3QLXV2VANCNFSM4IGP2CKQ>
> .
>

rniwa commented:
Great. Thanks for all your hard work, @rakina. I'm really excited that this feature is now fully spec'ed and implementable in WebKit. As a mater of fact, I've loaded the patch to implement this new behavior in https://bugs.webkit.org/show_bug.cgi?id=166484.

Hopefully @smaug---- can do another pass of the PR and we can call it good.
